File: app/src/main/java/me/saket/dank/ui/subreddit/uimodels/SubmissionItemDiffer.java
package me.saket.dank.ui.subreddit.uimodels;

import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import me.saket.dank.utils.SimpleDiffUtilsCallbacks;
import me.saket.dank.ui.subreddit.uimodels.SubredditScreenUiModel.SubmissionRowUiModel;

public class SubmissionItemDiffer extends SimpleDiffUtilsCallbacks<SubmissionRowUiModel> {

  public static SubmissionItemDiffer create(List<SubmissionRowUiModel> oldItems, List<SubmissionRowUiModel> newItems) {
    return new SubmissionItemDiffer(oldItems, newItems);
  }

  private SubmissionItemDiffer(List<SubmissionRowUiModel> oldItems, List<SubmissionRowUiModel> newItems) {
    super(oldItems, newItems);
  }

  @Override
  public boolean areItemsTheSame(SubmissionRowUiModel oldItem, SubmissionRowUiModel newItem) {
    return oldItem.adapterId() == newItem.adapterId();
  }

  @Override
  protected boolean areContentsTheSame(SubmissionRowUiModel oldItem, SubmissionRowUiModel newItem) {
    return oldItem.equals(newItem);
  }

  @Nullable
  @Override
  public Object getChangePayload(SubmissionRowUiModel oldItem, SubmissionRowUiModel newItem) {
    switch (oldItem.type()) {
      case SUBMISSION:
        SubredditSubmission.UiModel oldSubmission = (SubredditSubmission.UiModel) oldItem;
        SubredditSubmission.UiModel newSubmission = (SubredditSubmission.UiModel) newItem;

        List<SubredditSubmission.PartialChange> partialChanges = new ArrayList<>(3);
        if (oldSubmission.displayThumbnailOnLeftSide() != newSubmission.displayThumbnailOnLeftSide()) {
          partialChanges.add(SubredditSubmission.PartialChange.THUMBNAIL_POSITION);
        }
        if (!oldSubmission.title().equals(newSubmission.title())) {
          partialChanges.add(SubredditSubmission.PartialChange.TITLE);
        }
        if (!oldSubmission.byline().equals(newSubmission.byline())) {
          partialChanges.add(SubredditSubmission.PartialChange.BYLINE);
        }
        if (!oldSubmission.thumbnail().equals(newSubmission.thumbnail())) {
          partialChanges.add(SubredditSubmission.PartialChange.THUMBNAIL);
        }
        if (oldSubmission.isSaved() != newSubmission.isSaved()) {
          partialChanges.add(SubredditSubmission.PartialChange.SAVE_STATUS);
        }
        return partialChanges;

      case PAGINATION_FOOTER:
      case GESTURES_WALKTHROUGH:
        return super.getChangePayload(oldItem, newItem);

      default:
        throw new AssertionError();
    }
  }
}

File: app/src/main/java/me/saket/dank/ui/subreddit/uimodels/SubmissionItemDiffer.kt
package me.saket.dank.ui.subreddit.uimodels

import androidx.recyclerview.widget.DiffUtil
import me.saket.dank.ui.subreddit.uimodels.SubredditScreenUiModel.SubmissionRowUiModel

object SubmissionItemDiffer : DiffUtil.ItemCallback<SubmissionRowUiModel>() {

  override fun areItemsTheSame(oldItem: SubmissionRowUiModel, newItem: SubmissionRowUiModel): Boolean {
    return oldItem.adapterId() == newItem.adapterId()
  }

  override fun areContentsTheSame(oldItem: SubmissionRowUiModel, newItem: SubmissionRowUiModel): Boolean {
    return oldItem == newItem
  }

  override fun getChangePayload(oldItem: SubmissionRowUiModel, newItem: SubmissionRowUiModel): Any? {
    return when (oldItem.type()) {
      SubmissionRowUiModel.Type.SUBMISSION -> {
        val oldSubmission = oldItem as SubredditSubmission.UiModel
        val newSubmission = newItem as SubredditSubmission.UiModel

        val partialChanges = mutableListOf<SubredditSubmission.PartialChange>()
        if (oldSubmission.displayThumbnailOnLeftSide() != newSubmission.displayThumbnailOnLeftSide()) {
          partialChanges.add(SubredditSubmission.PartialChange.THUMBNAIL_POSITION)
        }
        if (oldSubmission.title() != newSubmission.title()) {
          partialChanges.add(SubredditSubmission.PartialChange.TITLE)
        }
        if (oldSubmission.byline() != newSubmission.byline()) {
          partialChanges.add(SubredditSubmission.PartialChange.BYLINE)
        }
        if (oldSubmission.thumbnail() != newSubmission.thumbnail()) {
          partialChanges.add(SubredditSubmission.PartialChange.THUMBNAIL)
        }
        if (oldSubmission.isSaved != newSubmission.isSaved) {
          partialChanges.add(SubredditSubmission.PartialChange.SAVE_STATUS)
        }
        if (oldSubmission.swipeActions() != newSubmission.swipeActions()) {
          partialChanges.add(SubredditSubmission.PartialChange.SWIPE_ACTIONS)
        }
        partialChanges
      }

      SubmissionRowUiModel.Type.PAGINATION_FOOTER,
      SubmissionRowUiModel.Type.GESTURES_WALKTHROUGH -> {
        super.getChangePayload(oldItem, newItem)
      }

      else -> throw AssertionError()
    }
  }
}
